CORE-8610 change cpk pk

conal-smith opened the pull request:
Read description given in API change first - https://github.com/corda/corda-api/pull/759

- Force uploading a CPK that has changed its liquibase changelog files will result in a new checksum so we no longer have to worry keeping an audit track of entity versions of changelogs in `cpk_db_change_log`.
- We drive rolling back of changelogs applied to a virtual node's vault by:
  - When we run the initial migrations we tag each changelog with the cpkFileChecksum that it came from.
  - If we have to roll back we list the checksums of all the changelogs that have been applied (query to liquibase's databasechangelog table) and query the `cpk_db_change_log` table to get the list of changelogs to roll back. The audit table is not involed.
- Audit table becomes an append only trail of the history of CPK related changes.

<details>
<summary>Expand here for more details...</summary>
Uploaded CPI with two CPKs:
	* cart (with cart changelog)
	* dog (with dog changelog)
![Image](https://user-images.githubusercontent.com/68279385/211372733-50310d43-7e85-40e3-a8cb-b56542e22cd3.png)

liquibase databasechangelog has these changelogs with their CPK file checksum as the tag.
![Image](https://user-images.githubusercontent.com/68279385/211372763-160a1411-4e4d-47e6-b505-9f1ceb4bdf56.png)

Which correspond to these CPKs:
![Image](https://user-images.githubusercontent.com/68279385/211372798-55dfdc3f-6ddd-4ac9-ae65-ff1c617c27dc.png)

This table shows the CPKs related to our test-cordapp_test-conal
![Image](https://user-images.githubusercontent.com/68279385/211372832-6bf07ae5-367b-4913-9826-11a3106d1085.png)

Now we've force uploaded the test-cordapp_test-conal with new CPKs:
![Image](https://user-images.githubusercontent.com/68279385/211372854-f9971db1-38be-4bbb-ab8f-fc5a45122ab2.png)
* Note how 40c (the shared CPK with cart changelog) is still associated with the CPI, and dogs (E87 the CPK) is no longer associated.

From a db changelog perspective, we have changelogs for each CPK:
![Image](https://user-images.githubusercontent.com/68279385/211372892-d1b084f6-bda3-4142-8b41-d0f18a491ab8.png)

A query at this stage to get all changelogs associated with the CPI would yield changelogs with cpkFileChecksum 40C, ED9 and 830 - corresponding to changelogs fish & owner (and their master changelog), cart (and its master changelog), but not dogs.
* this query will be used when resyncing the vault schema.

When we come to run the force-resync API, we query liquibase's databasechangelog table to get the list of tags (CPK file checksums) that have been applied, this yields list above, 40C and E87.

It then loads the db change logs for these checksums, rolls them back, and applies the new changelogs.

After resync, notice now the cart is rolled back, dog is rolled back, cart is reinstated, and fish table exists.
![Image](https://user-images.githubusercontent.com/68279385/211372960-fb4f7615-4262-4ae4-8a78-428b432513c9.png)

The logs read here:

```
2023-01-09 16:58:31.202 [qtp842884364-336] INFO  net.corda.httprpc.server.impl.context.ContextUtils.POST {http.method=POST, http.path=/api/v1/maintenance/virtualnode/FCE56A416484/vault-schema/force-resync, http.user=admin} - Servicing POST request to '/api/v1/maintenance/virtualnode/FCE56A416484/vault-schema/force-resync'
2023-01-09 16:58:31.203 [qtp842884364-336] INFO  net.corda.libs.virtualnode.maintenance.rpcops.impl.v1.VirtualNodeMaintenanceRPCOpsImpl {http.method=POST, http.path=/api/v1/maintenance/virtualnode/FCE56A416484/vault-schema/force-resync, http.user=admin} - Requesting synchronization of vault schema for virtual node 'FCE56A416484' with its current CPI.
2023-01-09 16:58:32.170 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor {} - CPK file checksums of currently applied changelogs on vault schema for virtual node FCE56A416484: [SHA-256:40C096983167D2ED37EBE20D8053A2127FDF83383E47E3604F6473230A1F3C43, SHA-256:E87F1AFE5AAF0C8B074989CCA47E0240C7D4DED662A8F8F5245A9FC766B1B9B1]
2023-01-09 16:58:32.171 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor {} - Virtual node 'FCE56A416484' attempting to roll back the following changelogs for CPK 'SHA-256:40C096983167D2ED37EBE20D8053A2127FDF83383E47E3604F6473230A1F3C43' [migration/cart-migration-v1.0.xml, migration/db.changelog-master.xml]
2023-01-09 16:58:32.215 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.database {} - Set default schema name to vnode_vault_fce56a416484
2023-01-09 16:58:32.270 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully acquired change log lock
2023-01-09 16:58:32.270 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.db.admin.impl.StreamResourceAccessor {} - Creating composite master changelog XML file master-changelog-0ff94924-8f8f-4840-aacf-739e3a56b828.xml with: [migration/db.changelog-master.xml]
2023-01-09 16:58:32.377 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - Reading from vnode_vault_fce56a416484.databasechangelog
2023-01-09 16:58:32.487 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully released change log lock
2023-01-09 16:58:32.487 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.db.admin.impl.LiquibaseSchemaMigratorImpl {} - cordacluster DB schema rollback complete
2023-01-09 16:58:32.487 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor {} - Migrations for CPK 'SHA-256:40C096983167D2ED37EBE20D8053A2127FDF83383E47E3604F6473230A1F3C43' successfully rolled back.
2023-01-09 16:58:32.487 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor {} - Virtual node 'FCE56A416484' attempting to roll back the following changelogs for CPK 'SHA-256:E87F1AFE5AAF0C8B074989CCA47E0240C7D4DED662A8F8F5245A9FC766B1B9B1' [migration/db.changelog-master.xml, migration/dogs-migration-v1.0.xml]
2023-01-09 16:58:32.535 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.database {} - Set default schema name to vnode_vault_fce56a416484
2023-01-09 16:58:32.599 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully acquired change log lock
2023-01-09 16:58:32.599 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.db.admin.impl.StreamResourceAccessor {} - Creating composite master changelog XML file master-changelog-e025a056-461b-4a2c-b20f-8c87423247de.xml with: [migration/db.changelog-master.xml]
2023-01-09 16:58:32.702 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - Reading from vnode_vault_fce56a416484.databasechangelog
2023-01-09 16:58:32.761 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully released change log lock
2023-01-09 16:58:32.762 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.db.admin.impl.LiquibaseSchemaMigratorImpl {} - cordacluster DB schema rollback complete
2023-01-09 16:58:32.762 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor {} - Migrations for CPK 'SHA-256:E87F1AFE5AAF0C8B074989CCA47E0240C7D4DED662A8F8F5245A9FC766B1B9B1' successfully rolled back.
2023-01-09 16:58:32.800 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor {} - Preparing to run 3 resync migrations for CPK 'SHA-256:ED998E047A916C23D69E332A04C6C3BC552B258BB23EBE345B677075078DF0EC'.
2023-01-09 16:58:32.815 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.database {} - Set default schema name to vnode_vault_fce56a416484
2023-01-09 16:58:32.817 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.db.admin.impl.LiquibaseSchemaMigratorImpl {} - Updating PostgreSQL 14.4 (Debian 14.4-1.pgdg110+1) DB Schema for cordacluster
2023-01-09 16:58:32.881 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully acquired change log lock
2023-01-09 16:58:32.882 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.db.admin.impl.StreamResourceAccessor {} - Creating composite master changelog XML file master-changelog-5dbc6d57-2d25-46ad-820a-726013ad7a5d.xml with: [migration/db.changelog-master.xml]
2023-01-09 16:58:32.954 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - Reading from vnode_vault_fce56a416484.databasechangelog
2023-01-09 16:58:33.038 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - Table fish created
2023-01-09 16:58:33.058 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - Primary key added to fish (id)
2023-01-09 16:58:33.066 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - ChangeSet migration/fish-migration-v1.0.xml::fish-migrations-v1.0::R3.Corda ran successfully in 52ms
2023-01-09 16:58:33.106 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - Table owner created
2023-01-09 16:58:33.126 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - Primary key added to owner (id)
2023-01-09 16:58:33.132 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - ChangeSet migration/owner-migration-v1.0.xml::owner-migrations-v1.0::R3.Corda ran successfully in 41ms
2023-01-09 16:58:33.153 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully released change log lock
2023-01-09 16:58:33.211 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully acquired change log lock
2023-01-09 16:58:33.318 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully released change log lock
2023-01-09 16:58:33.318 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.db.admin.impl.LiquibaseSchemaMigratorImpl {} - cordacluster DB schema update complete
2023-01-09 16:58:33.318 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor {} - Resync migrations for CPK 'SHA-256:ED998E047A916C23D69E332A04C6C3BC552B258BB23EBE345B677075078DF0EC' completed.
2023-01-09 16:58:33.355 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor {} - Preparing to run 2 resync migrations for CPK 'SHA-256:40C096983167D2ED37EBE20D8053A2127FDF83383E47E3604F6473230A1F3C43'.
2023-01-09 16:58:33.370 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.database {} - Set default schema name to vnode_vault_fce56a416484
2023-01-09 16:58:33.371 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.db.admin.impl.LiquibaseSchemaMigratorImpl {} - Updating PostgreSQL 14.4 (Debian 14.4-1.pgdg110+1) DB Schema for cordacluster
2023-01-09 16:58:33.424 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully acquired change log lock
2023-01-09 16:58:33.424 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.db.admin.impl.StreamResourceAccessor {} - Creating composite master changelog XML file master-changelog-07306df5-b249-4278-afa5-d3249a885408.xml with: [migration/db.changelog-master.xml]
2023-01-09 16:58:33.485 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - Reading from vnode_vault_fce56a416484.databasechangelog
2023-01-09 16:58:33.561 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - Table cart created
2023-01-09 16:58:33.580 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - Primary key added to cart (id)
2023-01-09 16:58:33.586 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.changelog {} - ChangeSet migration/cart-migration-v1.0.xml::cart-migration-v1.0::R3.Corda ran successfully in 47ms
2023-01-09 16:58:33.610 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully released change log lock
2023-01-09 16:58:33.667 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully acquired change log lock
2023-01-09 16:58:33.770 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  liquibase.lockservice {} - Successfully released change log lock
2023-01-09 16:58:33.770 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.db.admin.impl.LiquibaseSchemaMigratorImpl {} - cordacluster DB schema update complete
2023-01-09 16:58:33.770 [rpc subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor {} - Resync migrations for CPK 'SHA-256:40C096983167D2ED37EBE20D8053A2127FDF83383E47E3604F6473230A1F3C43' completed.
2023-01-09 16:58:34.092 [rpc response subscription thread virtual.node.management-virtual.node.creation.request] INFO  net.corda.messaging.publisher.CordaRPCSenderImpl {} - Response for request 1bd9c12e-4b7b-4349-a697-ff4bbcdccdd1 was received at 2023-01-09T16:58:33.774Z
```

After this, the databasechangelog for the vault looks as follows:
![Image](https://user-images.githubusercontent.com/68279385/211373036-d25f4a8f-2316-48f5-ab28-148b338f8d03.png)
* cart has correct migrations applied.
* Note - there is an existing bug where only one of the changelogs in the CPK is tagged, not all of them, this will be dealt with separately. For now this doesn't manifest as an issue because we only use this table to get the cpk file checksums of changelogs that were applied, we get ED99 so we know fish is in this CPK.

If we force upload the original CPI back again, we see the correct vault schema.
![Image](https://user-images.githubusercontent.com/68279385/211373211-c220993c-b656-496b-aa9e-e7d619d23de1.png)

</details>

driessamyn commented on components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt:
```diff
@@ -132,20 +124,19 @@ internal class DatabaseCpiPersistenceTest {
+        fileChecksum: SecureHash = newRandomSecureHash(),
+        name: String = UUID.randomUUID().toString(),
+        version: String = "cpk-version",
+        ssh: SecureHash = newRandomSecureHash()
```
I know the `ssh` name has been introduced elsewhere, but I would stay away from it as it's a well know abbreviation for `Secure Shell` and is a bit confusion. I think it's better to stick to `signerSummaryHash` consistently.

driessamyn commented on components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt:
```diff
@@ -256,48 +250,80 @@ internal class DatabaseCpiPersistenceTest {
+
+        val forceUploadedCpi = loadCpiDirectFromDatabase(updatedCpi)
+
+        // optimistic force increment + calling merge on this entity has incremented by 2
+        assertThat(forceUploadedCpi.entityVersion).isEqualTo(3)
```
Bit of a detail and probably not introduced in this PR, but does this need to increment by 2? That feels a bit wrong.

sprintengine replied:
Afaik first increment is caused by the optimistic force increment flag we set on the find query

https://github.com/corda/corda-runtime-os/blob/54f6cc660e3559a94bfb76807ebbba98b6a1764e/components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/database/DatabaseCpiPersistence.kt#L242-L250

Second increment is caused by calling merge on the CPI entity.

driessamyn replied:
And why do we set this, do you know? It feels like we should maybe always be incrementing it when we write, not read.

sprintengine replied:
https://github.com/corda/corda-runtime-os/commit/4072aa9e6b4f40a6f8dd4a5edcf4f82701f7998e

Looks like it was an attempt to bump CPI version when we force upload a CPI, I'm not sure if it's necessary since this does seem to already happen when calling merge, maybe there is an edge case it was covering?

driessamyn commented on components/chunking/chunk-db-write-impl/src/integrationTest/kotlin/net/corda/chunking/db/impl/tests/DatabaseCpiPersistenceTest.kt:
```diff
@@ -256,48 +250,80 @@ internal class DatabaseCpiPersistenceTest {
+        assertThat(forceUploadedCpk2.metadata.entityVersion).isEqualTo(0)
+
+//        assertChangeLogPersistedWithCpi(cpk1)
+//        assertChangeLogPersistedWithCpi(cpk2)
```
Can these be put back?

vkolomeyko commented on libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/CpiCpkEntity.kt:
```diff
@@ -23,22 +22,12 @@ import javax.persistence.Version
 @Table(name = "cpi_cpk", schema = DbSchema.CONFIG)
 data class CpiCpkEntity(
```
Referenced Jira also says: "Apply recent JPA guidance."
should we do this as part of this PR (e.g. eliminate `data class` on entities) or would you rather do it as a separate one?

I would probably go for the latter, as this PR is sufficiently big already.

sprintengine replied:
Thanks I meant to update the description, CORE-8744 covers this JPA guidance refactoring.

wzur-r3 commented on gradle.properties:
```diff
@@ -40,7 +40,7 @@ bouncycastleVersion=1.72
 # Change to 5.0.0.xx-SNAPSHOT to pick up maven local published copy
 #cordaApiVersion=5.0.0.xxx-SNAPSHOT
-cordaApiVersion=5.0.0.605-beta+
+cordaApiVersion=5.0.0.606-alpha-1673001400834
```
that is surely wrong

vkolomeyko replied:
I think the plan is to change it to `beta` version once associated `corda-api` PR is merged.

sprintengine replied:
Yes, we typically achieve a green build using the alpha artifact from corda-api first, I'll be changing this to 606-beta+ if corda-api gets approval

conalsmith-r3 replied:
Can you lift the restriction on this otherwise you become a dependency on this getting merged and the release of a corda-api version lock

driessamyn commented on components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterProcessor.kt:
```diff
@@ -286,14 +300,19 @@ internal class VirtualNodeWriterProcessor(
-                            changelogs
-                        )
-                    }
+                    // note cs - I noticed this function may kill the thread running liquibase API if it encounters an exception.
+                    //  I think this could be related to embedding creation of datasources inside entity manager transactions and then
+                    //  expecting the entity manager to roll back when liquibase throws. Liquibase handles its own transactionality (one
+                    //  transaction per changeset) and so we likely should not imply from the layout of this code that we get
+                    //  transactionality, and the code should likely be refactored.
+                    //  Example scenario: force upload a CPI with a CPK with a liquibase changelog that attempts to create a table that is
+                    //  already created, then call the resync function, will kill the db-worker and require restart of combined worker.
+                    //  Perhaps this refactoring can be handled in CORE-8744 and a suitable test created to handle liquibase exceptions
+                    //  when running resync migrations.
+                    runCpiResyncMigrations(
```
This is a good observation. It think we should actually create a bug JIRA for this. Maybe it makes sense to look at it alongside CORE-8744, but we need to be careful it does't fall through the cracks as it sounds like it could cause a worker to fail.

sprintengine commented on components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/database/DatabaseCpiPersistence.kt:
```diff
@@ -68,128 +58,137 @@ class DatabaseCpiPersistence(private val entityManagerFactory: EntityManagerFact
+        val relationshipsForExistingCpks = existingCpks.map { thisCpk ->
+            val cpkFileChecksum = thisCpk.metadata.fileChecksum.toString()
+            val cpkEntity = foundCpks[cpkFileChecksum]!!
+            CpiCpkEntity(
```
This is wrong, we should first try load CpiCpkEntity because it might exist, for example, if CPI force uploaded doesn't change a CPK it will already be associated with the CPI and merging this entity could fail with optimistic lock exception

sprintengine commented on components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterProcessor.kt:
```diff
@@ -286,14 +300,19 @@ internal class VirtualNodeWriterProcessor(
+                    //  Perhaps this refactoring can be handled in CORE-8744 and a suitable test created to handle liquibase exceptions
+                    //  when running resync migrations.
+                    runCpiResyncMigrations(
+                        dbConnectionManager.createDatasource(virtualNodeInfo.vaultDdlConnectionId!!),
```
This is missing a close of the datasource

sprintengine commented on components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterProcessor.kt:
```diff
@@ -276,23 +277,69 @@ internal class VirtualNodeWriterProcessor(
+                                    )
+                                )
+                            )
+                        }
```
This is all quite messy, but https://r3-cev.atlassian.net/browse/CORE-8744 will refactor this code into repositories, etc.

sprintengine commented on components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterProcessor.kt:
```diff
@@ -276,23 +277,69 @@ internal class VirtualNodeWriterProcessor(
+                    //  Perhaps this refactoring can be handled in CORE-8744 and a suitable test created to handle liquibase exceptions
+                    //  when running resync migrations.
+                    val changelogsToRun = getCurrentChangelogsForCpi(em, cpiMetadata.id).groupBy { it.id.cpkFileChecksum }
+                    changelogsToRun.forEach { (cpkFileChecksum, changelogsForThisCpk) ->
```
I found when there were multiple CPK's changelogs to update, there was a liquibase exception in the original function:

```
java.sql.SQLTransientConnectionException: HikariPool-22 - Connection is not available, request timed out after 30005ms.
	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:696) ~[?:?]
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:181) ~[?:?]
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:146) ~[?:?]
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100) ~[?:?]
	at net.corda.db.core.HikariDataSourceFactory$DataSourceWrapper.getConnection(HikariDataSourceFactory.kt) ~[?:?]
	at net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor.runCpiResyncMigrations(VirtualNodeWriterProcessor.kt:633) ~[?:?]
	at net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor.resetVirtualNodeDb(VirtualNodeWriterProcessor.kt:341) ~[?:?]
	at net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor.onNext(VirtualNodeWriterProcessor.kt:466) ~[?:?]
	at net.corda.virtualnode.write.db.impl.writer.VirtualNodeWriterProcessor.onNext(VirtualNodeWriterProcessor.kt:83) ~[?:?]
	at net.corda.messaging.subscription.RPCSubscriptionImpl.processRecords$lambda$6$lambda$5(RPCSubscriptionImpl.kt:196) ~[?:?]
	at io.micrometer.core.instrument.composite.CompositeTimer.recordCallable(CompositeTimer.java:129) ~[?:?]
	at net.corda.messaging.subscription.RPCSubscriptionImpl.processRecords(RPCSubscriptionImpl.kt:196) ~[?:?]
	at net.corda.messaging.subscription.RPCSubscriptionImpl.pollAndProcessRecords(RPCSubscriptionImpl.kt:115) ~[?:?]
	at net.corda.messaging.subscription.RPCSubscriptionImpl.createProducerConsumerAndStartPolling(RPCSubscriptionImpl.kt:106) ~[?:?]
	at net.corda.messaging.subscription.RPCSubscriptionImpl.runConsumeLoop(RPCSubscriptionImpl.kt:77) ~[?:?]
	at net.corda.messaging.subscription.RPCSubscriptionImpl.access$runConsumeLoop(RPCSubscriptionImpl.kt:34) ~[?:?]
	at net.corda.messaging.subscription.RPCSubscriptionImpl$threadLooper$1.invoke(RPCSubscriptionImpl.kt:48) ~[?:?]
	at net.corda.messaging.subscription.RPCSubscriptionImpl$threadLooper$1.invoke(RPCSubscriptionImpl.kt:48) ~[?:?]
	at net.corda.messaging.subscription.ThreadLooper.runConsumeLoop(ThreadLooper.kt:174) ~[?:?]
	at net.corda.messaging.subscription.ThreadLooper.access$runConsumeLoop(ThreadLooper.kt:27) ~[?:?]
	at net.corda.messaging.subscription.ThreadLooper$StoppableThread$start$1$1.invoke(ThreadLooper.kt:106) ~[?:?]
	at net.corda.messaging.subscription.ThreadLooper$StoppableThread$start$1$1.invoke(ThreadLooper.kt:106) ~[?:?]
	at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:30) ~[?:?]
```

I now run each CPK's migrations in a separate call to the liquibase API and close the connection after each run. https://r3-cev.atlassian.net/browse/CORE-8744 can likely improve this code.

sprintengine commented on components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeWriterProcessor.kt:
```diff
@@ -276,23 +277,69 @@ internal class VirtualNodeWriterProcessor(
+                                runCpkResyncMigrations(it, cpkFileChecksum, changelogsForThisCpk)
+                            }
+                        } catch (e: Exception) {
+                            logger.error(
```
Liquibase errors were causing the entire worker to go down. Catching here prevents this. Potentially we can log at a less severe level.

conal-smith committed:
CORE-8610 - Changing the primary key of CPK related entites to the file checksum
- Removal of most of rollback using audit logs, use cpk change logs instead
- add to test CPI multiple CPKs and add logic to rollback more than one CPK
- when extracting changelogs from CPI, no longer setting changesetId, this is removed.
- when running migrations for a virtual node, tag each changelog with the CPK file checksum it came from, this can be looked up when we need to roll back the change.
- improve the logic when persisting/updating cpis, including purging old cpiCpk relationships when no longer associated (but leaving the remaining CPK metadata, file and changelog).
- improved logging when it comes to rolling back changes, running migrations, changelog selection.
- added a DSL for easily creating CPI and CPK related entities

conal-smith committed:
CORE-8610 - use set when asserting contents of results for CPK change logs for a CPI

conal-smith committed:
CORE-8610 - add a test that force uploads and sync to a CPI with no changelogs

conal-smith committed:
CORE-8610 - log about changelogs that get persisted

conal-smith committed:
CORE-8610 - bump api

conal-smith committed:
CORE-8610 - fix compilation

conal-smith committed:
CORE-8610 bump api vers

conal-smith committed:
CORE-8610 bump api vers

driessamyn reviewed (approved):
LGTM

ronanbrowne reviewed (approved):
Lgtm - build

mag1c-48 reviewed (approved):
LGTM - Build
